### PR TITLE
Implemented numpy setxor1d

### PR DESCRIPTION
**Pull request by joschkabraun**
This is an implementation of [numpy.setxor1d](https://docs.scipy.org/doc/numpy/reference/generated/numpy.setxor1d.html) from issue #70 (and #2078). It follows the same structure as the original implementation with changes for Jax compatibility.

np.setxor1d is intrinsically hard to make jit-compatible since it depends on np.unique and its output shape depends on the values in the array. This problem will be resolved as soon as np.unique is jit-compaitible (#2760). 

The test check all one dimensional arrays and number dtypes which are not complex (as by #2760, unique does not work with complex dtypes). The test does not compile with jit since unique is not compatible with it.

### CODE REVIEW

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3174,6 +3174,28 @@ def unique(ar, return_index=False, return_inverse=False,
...
         "np.unique is not implemented for the axis argument")
 
+  
+@_wraps(onp.setxor1d)
+def setxor1d(ar1, ar2, assume_unique=False):
+  if not assume_unique:
+    ar1 = unique(ar1)
+    ar2 = unique(ar2)
+  else:
+    if ar1.ndim == 0:
```
Comment (jekbradbury) [inline jax/numpy/lax_numpy.py:3184]: This will fail if an input is a Python scalar rather than a NumPy/JAX scalar.

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -3174,6 +3174,28 @@ def unique(ar, return_index=False, return_inverse=False,
...
+      ar1 = array([ar1])
+    if ar2.ndim == 0:
+      ar2 = array([ar2])
+  
+  aux = concatenate((ar1, ar2))
+  if aux.size == 0:
+      return aux
+
+  aux.sort()
```
Comment (jekbradbury) [inline jax/numpy/lax_numpy.py:3193]: This relies on the in-place behavior of NumPy `sort`; you want to use something like `aux = sort(aux)` for JAX compatibility.

**Review by jekbradbury — CHANGES_REQUESTED** [review]

**Comment by joschkabraun** [comment]
Okay, thank you for your feedback! I’m currently working on it and will finish it until the weekend!

**Comment by joschkabraun** [comment]
The tests are running smooth on my local machine but I am getting errors here which seem to be outside of my code. Can you help me with them?

**Comment by jakevdp** [comment]
As with #2941, you need to rebase on master, fix tests locally, and then force push to update this PR. Please let me know if you need help with that process.

**Comment by jakevdp** [comment]
``rng_factory`` now requires ``self.rng()`` as the first argument - take a look at how other test routines use it.

**Comment by joschkabraun** [comment]
Okay, thank you!

**Comment by joschkabraun** [comment]
Running the tests here gives me the error "NotImplementedError: Numpy function <function setxor1d at 0x7fae7fd63ea0> not yet implemented". Do you know how that can happen?

**Comment by jakevdp** [comment]
> Running the tests here gives me the error "NotImplementedError: Numpy function <function setxor1d at 0x7fae7fd63ea0> not yet implemented"

You'll need to add the function to the top-level imports here: https://github.com/google/jax/blob/e48a4e012bd253428e3bb0dd4ac5b163439a5270/jax/numpy/__init__.py#L20-L59

**Comment by joschkabraun** [comment]
Okay, thank you!

**Comment by jakevdp** [comment]
Looks like the remaining error is a dtype mismatch in x64 mode. I'd suggest passing ``check_dtypes=False`` in the test function.

**Comment by joschkabraun** [comment]
Okay, thank you!

**Comment by jakevdp** [comment]
Looks good - please also remove the two ``.orig`` files that were added along the way.

**Comment by joschkabraun** [comment]
Okay, this has happened now!

File: jax/numpy/lax_numpy.py [hunk 3]
```diff
@@ -3320,6 +3319,24 @@ def unique(ar, return_index=False, return_inverse=False,
...
         "np.unique is not implemented for the axis argument")
 
+@_wraps(np.setxor1d)
+def setxor1d(ar1, ar2, assume_unique=False):
+  if not assume_unique:
+    ar1 = unique(ar1)
+    ar2 = unique(ar2)
+  else:
+    ar1 = asarray(ar1)
+    ar2 = asarray(ar2)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:3329]: We should not call `asarray`; valid inputs will be properly handled by `concatenate`.

File: jax/numpy/lax_numpy.py [hunk 4]
```diff
@@ -3320,6 +3319,24 @@ def unique(ar, return_index=False, return_inverse=False,
...
+  
+  aux = concatenate((ar1, ar2))
+  if aux.size == 0:
+      return aux
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:3333]: Use two-space indent to match the rest of the file.

File: tests/lax_numpy_test.py [hunk 5]
```diff
@@ -1262,6 +1262,23 @@ def testUnique(self, shape, dtype, return_index, return_inverse, return_counts):
...
+       "dtype2": dtype2}
+      for dtype1 in filter(lambda dt: dt != jnp.bfloat16, default_dtypes)
+      for dtype2 in filter(lambda dt: dt != jnp.bfloat16, default_dtypes)
+      for shape1 in (one_dim_array_shapes + [jtu.NUMPY_SCALAR_SHAPE])
+      for shape2 in (one_dim_array_shapes + [jtu.NUMPY_SCALAR_SHAPE])))
+  def testSetxor1d(self, shape1, dtype1, shape2, dtype2):
+    rng = jtu.rand_default(self.rng())
+    args_maker = lambda: [rng(shape1, dtype1), rng(shape2, dtype2)]
+    np_fun = lambda ar1, ar2: np.setxor1d(ar1, ar2)
+    jnp_fun = lambda ar1, ar2: jnp.setxor1d(ar1, ar2)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1279]: Since there's no complilation test, we can simplify to
```
np_fun = np.setxor1d
jnp_fun = jnp.setxor1d
```

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Comment by joschkabraun** [comment]
Okay, incorporated changes!

**Comment by jakevdp** [comment]
I just looked a bit closer at the numpy implementation: this seems to be essentially a line-for-line copy of the `setxor1d` code in numpy. That's fine, but for licensing purposes we'll need to move the implementation into ``jax/third_party/numpy`` before we can merge it.

**Comment by joschkabraun** [comment]
Okay, that makes sense! Should I create a file `lax_numpy.py` similar to the structure in `jax/numpy` as there is only `linalg.py` in `jax/third_party/numpy`?

**Comment by jakevdp** [comment]
Sure, you can create `jax/third_party/numpy/lax_numpy.py`, reflecting the structure of the main package.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Looks great! The last thing we need is a rebase against master to fix the conflicts that have cropped up over the last few days. Please let me know if you need tips on how to do that.

**Comment by jakevdp** [comment]
Hi – looks like something is strange with the rebase. If you need help getting your branch back in order, please let me know what commands you ran.

**Comment by jakevdp** [comment]
Since there has been no action here for a while, I'm going to close this. Feel free to re-open if you'd like to re-visit this contribution! Thanks

**Pull request closed without merging**